homeassistant/components/upnp.py

```diff
@@ -1,32 +1,54 @@
 """
-This module will attempt to open a port in your router for Home Assistant.
+Will open a port in your router for Home Assistant and provide statistics.
 
 For more details about this component, please refer to the documentation at
 https://home-assistant.io/components/upnp/
 """
+import asyncio
 import logging
 from urllib.parse import urlsplit
 
 import voluptuous as vol
 
 from homeassistant.const import (EVENT_HOMEASSISTANT_STOP)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+
+REQUIREMENTS = ['miniupnpc==1.9']
 
 _LOGGER = logging.getLogger(__name__)
 
 DEPENDENCIES = ['api']
 DOMAIN = 'upnp'
 
+DATA_UPNP = 'UPNP'
+
+CONF_ENABLE_PORT_MAPPING = 'port_mapping'
+CONF_UNITS = 'unit'
+
+UNITS = {
+    "Bytes": 1,
+    "KBytes": 1024,
+    "MBytes": 1024**2,
+    "GBytes": 1024**3,
+}
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({}),
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_ENABLE_PORT_MAPPING, default=True): cv.boolean,
+        vol.Optional(CONF_UNITS, default="MBytes"): vol.In(UNITS),
+    }),
 }, extra=vol.ALLOW_EXTRA)
 
 
 # pylint: disable=import-error, no-member, broad-except
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Register a port mapping for Home Assistant via UPnP."""
     import miniupnpc
 
     upnp = miniupnpc.UPnP()
+    hass.data[DATA_UPNP] = upnp
 
     upnp.discoverdelay = 200
     upnp.discover()
```

Comment:
Assigning to data doesn't do I/O. This comment was about `upnp.discover()` call. Since you changed it back to no longer be a coroutine this comment can be ignored.